design/pipeline-artifacts.md

```diff
@@ -37,23 +39,80 @@ Artifacts are distinct from other `resources` types, including `containers`, `re
+#### YAML / unified pipelines
+Artifacts are downloaded to the pipeline's workspace by default. We introduce a new variable for unified pipelines: `$(Pipeline.Workspace)`. `Pipeline.Workspace` is one level up from `$(System.DefaultWorkingDirectory)`; on hosted, it corresponds to `c:\agent\_work\1\`.
+
+Each artifact is given its own directory e.g. `$(Pipeline.Workspace)\myartifact` for an artifact named `myartifact`. Artifacts coming from other pipelines are each given one directory per pipeline e.g. `$(Pipeline.Workspace)\some-other-pipeline\someartifact` for the `someartifact` artifact of the `some-other-pipeline` pipeline. The pipeline portion of the path comes from the ID of the pipeline's `resource` in YAML.
+
+Example:
+```yaml
+resources:
+  pipelines:
+  - pipeline: mypipe
+
+jobs:
+- job: makeartifact
+  steps:
+  - script: ./build.sh
+  - upload: outputs/**/*
+
+- job: use1artifact
+  dependsOn: makeartifact
+  steps:
+  # by default, a download step is injected -- see later section
+  - script: ls $(Pipeline.Workspace)
+    # this listing shows a folder called `makeartifact`, since the prior job didn't specify a name
+
+- job: use2artifacts
+  dependsOn: makeartifact
+  steps:
+  - download: mypipe   # downloads all artifacts from `mypipe`
+  - download: current  # must include this, since by including a download step, we don't get automatic behavior anymore
+  - script: ls $(Pipeline.Workspace)
+    # this listing shows two folders: `makeartifact` and `mypipe`
+```
+
+#### Controlling download path
+
+If the `path` key is specified, it's a relative path from `$(Pipeline.Workspace)`. Directory names are not automatically injected by the pipeline anymore (but of course, directories present in the artifact itself are still used).
+
+Example:
+```yaml
+jobs:
+- job: makeartifact
+  steps:
+  - script: ./build.sh
+  - upload: outputs/**/*
+
+- job: useartifact
+  dependsOn: makeartifact
+  steps:
+  - download: current
+    path: foo
+  - script: ls $(Pipeline.Workspace)
+    # listing shows one folder, "foo"
+```
+
+#### Build and RM classic pipelines
+No change to current behavior. Artifacts are downloaded to `$(System.DefaultWorkingDirectory)`, which is the sources folder on Build and the artifacts folder on RM.
+
+**Note**: this means we need a new job message type "Pipeline" in addition to the existing "Build" and "Release".
 
 ### Schema
 
 ```yaml
-- download: string # identifier for the pipeline resource from which to download artifacts, optional; "current" means the current pipeline
-  artifact: string # identifier for the artifact to download; required
+- download: string # identifier for the pipeline resource from which to download artifacts, optional; "current" means the current pipeline, blank means all available pipelines (including current)
+  artifact: string # identifier for the artifact to download; optional
```

Comment:
I'm not opposed to removing this restriction, but one download keyword should map to exactly one task instance.

We need to build and leverage the same infrastructure (as much as possible), that we think will enable third party extensibility in the future.

Reply:
And the rules for how the directory structure is created, should be easy for customers to understand.

It would be good to explicitly write the details. And scrutinize whether the explanation is easy to understand.

Think through all cases. For example, explicit download for external pipeline, no artifact specified. A table might help think through the cases:

| Implicit/Explicit? | Pipeline has multiple artifacts? | Artifact specified? | Path specified? |
| --- | --- | --- | --- |
| Implicit | No | n/a | n/a |
| Implicit | Yes | n/a | n/a |
| Explicit | No | No | No |
| Explicit | No | No | Yes |
| Explicit | No | Yes | No |
| Explicit | No | Yes | Yes |
| Explicit | Yes | No | No |
| Explicit | Yes | No | Yes |
| Explicit | Yes | Yes | No |
| Explicit | Yes | Yes | Yes |

Reply:
...and whether the pipeline is `current` or an external pipeline should also be included in the matrix

Reply:
This is really good feedback - thank you. I'll steal your table as a starting point.